### PR TITLE
interp: add debug interface

### LINKED ISSUE #1188: Discussion: Interactive debugger interface
### Proposal

Discuss introducing a mechanism for interactively debugging Yaegi scripts. For example, some `script.go`, I want some interface parallel to `interp.EvalPath("script.go")` that provides a way to step through the Go statements in `script.go`.

I am happy to develop the debugger myself, as part of this repo, or another traefik project, or a completely separate project. I would do this as a [Debug Adapter Protocol](https://microsoft.github.io/debug-adapter-protocol/) implementation.

### Background

I created the [Go Notebook Kernel](https://marketplace.visualstudio.com/items?itemName=ethan-reesor.go-notebook-kernel) extension for Visual Studio Code. I would like to add the capability to debug Go notebook cells. Since the notebook kernel is driven by Yaegi, this would require interfaces for debugging Interpreter.Eval, so that I could implement a debug adapter for it. I have previously developed the [VSCode Byebug Debugger](https://marketplace.visualstudio.com/items?itemName=ethan-reesor.vscode-byebug), which relies on my implementation of the debug adapter protocol for Byebug.

### Workarounds

N/A. Other than debugging Yaegi itself, there's no way to debug a Yaegi script.

### PR DESCRIPTION
**Pull request by firelizzard18**
Adds an interface to `interp` that can be used to build an interactive debugger, such as those used by IDEs.

Closes #1188

All basic debugger features work, with one exception: breakpoints in some locations don't work, due to `setExec` creating a temporary `bltn` (https://github.com/traefik/yaegi/issues/1188#issuecomment-886107905).

Example, using a Debug Adapter implementation with VSCode:

image

### CODE REVIEW

File: interp/interp.go [hunk 1]
```diff
@@ -27,30 +25,46 @@ import (
...
+	scope     *scope         // frame scope
+	action    action         // action
+	exec      bltn           // generated function to execute
+	gen       bltnGenerator  // generator function to produce above bltn
+	val       interface{}    // static generic value (CFG execution)
+	rval      reflect.Value  // reflection value to let runtime access interpreter (CFG)
+	ident     string         // set if node is a var or func
+	breakPos  bool           // breakpoint set by position
+	breakCall bool           // breakpoint set on function call
+	prog      *Program       // program this is part of, only stored on the root node
```
[the remark this answers is not in the document]
Comment (firelizzard18) [reply interp/interp.go]: I've moved all the debug stuff to a single struct pointer field on `node`

**Review by mvertes — APPROVED** [review]

**Comment by mvertes** [comment]
@firelizzard18 about the rest of your work (the full yaegi VS code debug adapter), and to follow on your suggestion, we are thinking about setting a  dedicated GitHub repo / organisation to let community members, and not only Traefik Labs employees, to take part to the governance and maintenance of this kind of projects. This is a company wide reflection, and not only for your project, it may take a little bit more time to finalize, thanks for your patience.

**Comment by mvertes** [comment]
@firelizzard18 we have created a GitHub organization, empty at the moment, https://github.com/traefik-contrib, as discussed previously, in order to welcome and host interesting community projects around traefik and yaegi. If you are ok, we send you an invitation to be part of this organization, and so you can host your interactive debugger project here. For information, as a start. there is also a couple of traefik plugins from community that we would like to host as well.

We have also opened the GitHub [discussions](https://github.com/traefik/yaegi/discussions) in the yaegi repo, so we have a place to discuss.

**Comment by firelizzard18** [comment]
@mvertes Sounds good :+1:

I recently started working full-time, so I'm not sure how much time I'll be able to devote to Yaegi, but I can definitely clean up my debugger projects and push them to traefik-contrib.

**Comment by ldez** [comment]
@firelizzard18 I send you an invitation to the https://github.com/traefik-contrib organization. Ping me if you need help or information.

**Comment by firelizzard18** [comment]
@ldez @mvertes I created two repositories, for the VSCode extension and the debug adapter. I still need to clean them up, but my day job has been eating my time and energy, so I went ahead and pushed what I have.

**Pull request merged**